src/sst/core/model/python/pymodel_unitalgebra.cc

```diff
@@ -0,0 +1,389 @@
+        break;
+    case Py_GT:
+        if ( *(self_ua->obj) > *(other_ua->obj) ) return Py_True;
+        else return Py_False;
+        break;
+    case Py_GE:
+        if ( *(self_ua->obj) >= *(other_ua->obj) ) return Py_True;
+        else return Py_False;
+        break;
+    case Py_EQ:
+    case Py_NE:
+    default:
+        return Py_NotImplemented;
+        break;
+    }
+    return Py_NotImplemented;
+}
+
+
+// Numerical functions
+static PyObject* unitAlgebraAddIP(PyObject *self, PyObject *other)
+{
+    UnitAlgebraPy_t *self_ua = (UnitAlgebraPy_t*)self;
+    UnitAlgebraPy_t *other_ua = (UnitAlgebraPy_t*)other;
+    *(self_ua->obj) += *(other_ua->obj);
+
+    PyObject *argList = Py_BuildValue("(O)", self);
+    PyObject* ret = PyObject_CallObject((PyObject*)&PyModel_UnitAlgebraType, argList);    
+    Py_DECREF(argList);
+    
+    return ret;
+}
+
+static PyObject* unitAlgebraAdd (PyObject *self, PyObject *other)
+{
+    PyObject *argList = Py_BuildValue("(O)", self);
+    PyObject* ret = PyObject_CallObject((PyObject*)&PyModel_UnitAlgebraType, argList);    
+    Py_DECREF(argList);
+    
+    UnitAlgebraPy_t *ret_ua = (UnitAlgebraPy_t*)ret;
+    UnitAlgebraPy_t *other_ua = (UnitAlgebraPy_t*)other;
+    *(ret_ua->obj) += *(other_ua->obj);
+    return ret;
+}
+    
+    
+static PyObject* unitAlgebraSubIP(PyObject *self, PyObject *other)
+{
+    UnitAlgebraPy_t *self_ua = (UnitAlgebraPy_t*)self;
+    UnitAlgebraPy_t *other_ua = (UnitAlgebraPy_t*)other;
+    *(self_ua->obj) -= *(other_ua->obj);
+
+    PyObject *argList = Py_BuildValue("(O)", self);
+    PyObject* ret = PyObject_CallObject((PyObject*)&PyModel_UnitAlgebraType, argList);    
+    Py_DECREF(argList);
+    
+    return ret;
+}
+
+static PyObject* unitAlgebraSub (PyObject *self, PyObject *other)
+{
+    PyObject *argList = Py_BuildValue("(O)", self);
+    PyObject* ret = PyObject_CallObject((PyObject*)&PyModel_UnitAlgebraType, argList);    
+    Py_DECREF(argList);
+
+    UnitAlgebraPy_t *ret_ua = (UnitAlgebraPy_t*)ret;
+    UnitAlgebraPy_t *other_ua = (UnitAlgebraPy_t*)other;
+    *(ret_ua->obj) -= *(other_ua->obj);
+    return ret;
+}
+    
+static PyObject* unitAlgebraMulIP(PyObject *self, PyObject *other)
+{
+    UnitAlgebraPy_t *self_ua = (UnitAlgebraPy_t*)self;
+    UnitAlgebraPy_t *other_ua = (UnitAlgebraPy_t*)other;
+    *(self_ua->obj) *= *(other_ua->obj);
+
+    PyObject *argList = Py_BuildValue("(O)", self);
+    PyObject* ret = PyObject_CallObject((PyObject*)&PyModel_UnitAlgebraType, argList);    
+    Py_DECREF(argList);
+    
+    return ret;
+}
+
+static PyObject* unitAlgebraMul (PyObject *self, PyObject *other)
+{
+    PyObject *argList = Py_BuildValue("(O)", self);
+    PyObject* ret = PyObject_CallObject((PyObject*)&PyModel_UnitAlgebraType, argList);    
+    Py_DECREF(argList);
+
+    UnitAlgebraPy_t *ret_ua = (UnitAlgebraPy_t*)ret;
+    UnitAlgebraPy_t *other_ua = (UnitAlgebraPy_t*)other;
+    *(ret_ua->obj) *= *(other_ua->obj);
```

Comment:
If it is a shallow copy, would this be incorrect behavior? Shouldn't the original unit algebra values remain unmodified?

Even if a deep copy of the Python object, since we are using UnitAlgebra pointers, we are modifying the original UnitAlgebra.

```
a = b + c
``` 
should leave b and c unmodified, but this code would modify b

Reply:
Resolve with change to direct object rather than pointer.